### PR TITLE
Sameelement behaves like an and with extra constraints.

### PR DESCRIPTION
**Pull request by baldersheim**
So it should behave the sameway during fetchPostings too.

@havardpe @geirst PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/queryeval/executeinfo.h [hunk 1]
```diff
@@ -22,13 +22,10 @@ class ExecuteInfo {
...
-        return create(strict, 1.0F, doom);
-    }
     static ExecuteInfo create(bool strict, float hitRate, const vespalib::Doom * doom) noexcept {
         return {strict, hitRate, doom};
     }
-    static ExecuteInfo create(bool strict) noexcept {
+    static ExecuteInfo createForTest(bool strict) noexcept {
         return create(strict, 1.0F);
     }
     static ExecuteInfo create(bool strict, float hitRate) noexcept {
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/executeinfo.h:31]: maybe this also is 'for test'?

File: searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp [hunk 2]
```diff
@@ -56,8 +56,15 @@ SameElementBlueprint::optimize_self()
 void
 SameElementBlueprint::fetchPostings(const ExecuteInfo &execInfo)
 {
-    for (size_t i = 0; i < _terms.size(); ++i) {
-        _terms[i]->fetchPostings(ExecuteInfo::create(execInfo.isStrict() && (i == 0), execInfo));
+    if (_terms.empty()) return;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp:59]: I do not think 2x code is worth the one call to hit_ratio

**Review by havardpe — DISMISSED** [review]

File: searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp [hunk 3]
```diff
@@ -56,8 +56,15 @@ SameElementBlueprint::optimize_self()
...
+    _terms[0]->fetchPostings(execInfo);
+    double hit_rate = execInfo.hitRate();
+    const Blueprint * prev = _terms[0].get();
+    for (size_t i = 1; i < _terms.size(); ++i) {
+        hit_rate = hit_rate * prev->hit_ratio();
+        Blueprint * current = _terms[i].get();
+        current->fetchPostings(ExecuteInfo::create(false, execInfo));
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp:66]: Shouldn't `hit_rate` be used here?

File: searchlib/src/vespa/searchlib/queryeval/executeinfo.h (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/executeinfo.h]: It is, fixed.

File: searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp (hunk 3 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp]: Yes, indeed :)

File: searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp:59]: I did it for readability. I find it better to handle the first outside when it is different than the next. Here it is strictness that differs.

Comment (havardpe) [reply searchlib/src/vespa/searchlib/queryeval/same_element_blueprint.cpp:59]: I see, but you could still get rid of the current/prev shuffle if you called hit_ratio one extra time...

**Review by geirst — DISMISSED** [review]
👍

**Pull request merged**